### ISSUE TITLE
Assoc returning the dict

### ISSUE THREAD

**Issue by laarmen** [hylang/hy#238]
Would it be possible to have assoc return the modified dict instead of nil?

It would make dealing with dictionaries in Hy a bit easier, IMHO

**Comment by jmmk** [comment]
+1 on this. The current semantics don't make sense having an assoc form that returns None.

**Comment by emidln** [comment]
I'd prefer to see a `set!` that uses __setitem__ under the hood and make assoc return the modified dict.

**Comment by chr15m** [comment]
+1

**Comment by chr15m** [comment]
Maybe I am missing something but `(assoc foo bar baz)` could instead translate to:

`foo = dict(foo, **{bar: baz})`

http://stackoverflow.com/a/1453013/2131094

Hm, I should PR shouldn't I.

**Comment by kirbyfan64** [comment]
@chr15m Wouldn't that be slower, though? You're copying the entire dictionary to update one key.

Hy has the ability to implement this the correct way, similarly to `setv` and the `try` expression.

**Comment by chr15m** [comment]
Just tested - yes it is nearly 1000 times slower on my machine.

**Comment by chr15m** [comment]
![scream](https://cloud.githubusercontent.com/assets/67130/8084775/690af748-0fbf-11e5-8268-c883ea721b78.png)

**Comment by cwebber** [comment]
So, assoc is useful in a lisp using linked lists because it usually works with alists, which, while not the most efficient structures in the world, are (assuming nobody does any nasty stuff and mutate the cons cells), mostly immutable mapping structures.  Of course, sometimes people want more efficient variants... in Guile there are vlists, and there's a whole set of [purely functional data structures](https://github.com/ijp/pfds) and [fectors](https://github.com/ijp/fectors) available.  Which is great.  Often these reuse the name "assoc" in some way, but are purely functional!

So my suggestion is, instead of abusing assoc to try to fake immutability through having nice new structures returned on totally mutable structures by *copying* the dicts (yikes!), Hy should add nice sugar around existing immutable datastructures for Python... maybe [pyrsistent](https://github.com/tobgu/pyrsistent)?

**Comment by rcarmo** [comment]
FWIW, I think bolting in immutable data structures is another matter altogether. I would be very happy if assoc "just worked" and returned _something_ besides None/nil even if it mutated the dict. I know I'm using a Python data structure...

**Comment by chr15m** [comment]
If I introduced an `assoc!` which changes the dict and also returns it would that tiny PR be accepted?

**Comment by zackmdavis** [comment]
@chr15m I'd be in favor of such a function, but I can imagine the name `assoc!` being confusing (because the presence of both `assoc` and `assoc!` in the language might lead many to expect that `assoc` returns a new dict and doesn't modify the existing one, almost the opposite of what we actually want to convey). I'm not sure what alternative name to suggest at the moment.

**Comment by chr15m** [comment]
Good point.

The immediate situation is that the current `assoc` doesn't do what anybody expects. In my opinion the name of the current `assoc` should be something else (maybe `fast-assoc` or `set-key` or something better), and a new `assoc` that is slow but does what people expect should exist and have documentation that says why it is slower than the faster alternative. This gives the user a version of `assoc` that works as everyone expects it to, and also an alternative when they need a performant way to set dict key-vals inside e.g. a tight loop or whatever.

tl;dr: rename existing `assoc` and create a slower `assoc` that does what everybody expects.

**Comment by gilch** [comment]
How about compiling `assoc` to something more like this?
```Python
# current assoc
spam[i] = y

# proposed assoc
(spam.__setitem__(i, y) or spam)

# or even like this?
(spam.__setitem__(i,y), spam)[0]
```
Which of the two we use depends on if we want to allow custom `__setitem__` implementations on user classes to return something.

If we want the old behavior, for performance reasons, we rename the current `assoc` to something more intuitive, like `setitem` or `set-item`.

**Comment by zackmdavis** [comment]
@gilch, I like that idea! 

If `assoc` were to be renamed, I suggest `assoc*` (with an asterisk) as the new name, in analogy to how Hy and Clojure define the more-commonly-used `for` in terms of the more-fundamental `for*`.

**Comment by gilch** [comment]
A possible issue with my approach:
```Python
>>> spam = [1,2,3]
[1, 2, 3]
>>> def getspam():
...   print("side effect!")
...   return spam
...
>>> getspam().__setitem__(0,"zero") or getspam()
side effect!
side effect!
['zero', 2, 3]
```
I'm still not familiar enough with Hy's AST manipulation approach to know if this is easily avoidable.
The obvious way around this problem in handwritten pure Python has the overhead of a function call:
```Python
# (assoc (getspam) 0 "zero"))  ; compiles to:
(lambda x: x.__setitem(0, "zero))(getspam())
````
But this doesn't appear to work automatically, since then the `x` is not gensym'd and could shadow something important:
```Python
# (def x "something important")
# (assoc (getspam) 1 x)  ; compiles to:
x = "something important"
(lambda x: x.__setitem(0, x))(getspam())  # probably not what you meant.
```

I would also like to point out that Python already has an `operator.setitem` function. We don't need a new one, do we? Maybe we should import it by default though.

**Comment by gilch** [comment]
On second thought this should work:
```Python
x = "important"
(lambda x,i,y:x.__setitem__(i, y) or x)(getspam(), 0, "zero")
```
I noticed `assoc` is not considered a function now:
```
=> assoc
assoc
Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'assoc' is not defined
```
So there's a possible performance reason to duplicate the `operator.setitem` behavior as a macro.

**Comment by kirbyfan64** [comment]
I think I have a *really* simple solution. Basically, this:

```hy
(print (assoc d 1 2))
```

turns into:

```python
d[1] = 2
print(d)
```

and this:

```hy
(print (assoc (f) 1 2))
```

turns into:

```python
_hy_anon_var_x = f()
_hy_anon_var_x[1] = 2
print(_hy_anon_var_x)
```

**Comment by kirbyfan64** [comment]
After the discussion in #1246 and #1257, I'm going to close this. I don't see it happening now that setv also returns None.
